contracts/zero-ex/contracts/src/transformers/bridges/mixins/MixinFirebird.sol

```diff
@@ -0,0 +1,104 @@
+// SPDX-License-Identifier: Apache-2.0
+/*
+
+  Copyright 2021 ZeroEx Intl.
+
+  Licensed under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License.
+  You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+
+*/
+
+pragma solidity ^0.6.5;
+pragma experimental ABIEncoderV2;
+
+import "@0x/contracts-erc20/contracts/src/v06/LibERC20TokenV06.sol";
+import "@0x/contracts-erc20/contracts/src/v06/IERC20TokenV06.sol";
+import "../IBridgeAdapter.sol";
+
+/*
+    Firebird
+*/
+interface FirebirdRouter {
+
+    /// @dev Swaps an exact amount of input tokens for as many output tokens as possible, along the route determined by the path.
+    ///      The first element of path is the input token, the last is the output token, and any intermediate elements represent
+    ///      intermediate pairs to trade through (if, for example, a direct pair does not exist).
+    /// @param sellToken The address of the token to sell.
+    /// @param buyToken The address of the token to buy.
+    /// @param amountIn The amount of input tokens to send.
+    /// @param amountOutMin The minimum amount of output tokens that must be received for the transaction not to revert.
+    /// @param path Single element array with the address of the pool for sellToken and buyToken.
+    /// @param to Recipient of the output tokens.
+    /// @param deadline Unix timestamp after which the transaction will revert.
+    /// @return amounts The input token amount and all subsequent output token amounts.
+    function swapExactTokensForTokens(
+        IERC20TokenV06 sellToken,
+        IERC20TokenV06 buyToken,
+        uint amountIn,
+        uint amountOutMin,
+        IERC20TokenV06[] calldata path,
+        address to,
+        uint deadline
+    ) external returns (uint[] memory amounts);
+}
+
+contract MixinFirebird {
+
+    using LibERC20TokenV06 for IERC20TokenV06;
+
+    struct FirebirdBridgeData {
+        address router;
+        address pool;
+    }
+
+    function _tradeFirebird(
+        IERC20TokenV06 sellToken,
+        IERC20TokenV06 buyToken,
+        uint256 sellAmount,
+        bytes memory bridgeData
+    )
+        internal
+        returns (uint256 boughtAmount)
+    {
+        FirebirdBridgeData memory data = abi.decode(bridgeData, (FirebirdBridgeData));
+        FirebirdRouter router = FirebirdRouter(data.router);
+
+        IERC20TokenV06 pool;
+        address _pool = data.pool;
+        {
+            assembly {pool := _pool}
+        }
+        IERC20TokenV06[] memory path = new IERC20TokenV06[](1);
+        path[0] = pool;
```

Comment:
Happy for feedback if there is a better way to turn `FirebirdBridgeData.pool` of type `address` to a single element array of type `IERC20TokenV06`.

Reply:
I believe this "pattern" was only done in UniswapV3 because Solidity probably freaked out about decoding into an array of `IERC20TokenV06`

In this case they're not ERC20Tokens but pool addresses, so I would make the following changes.

```
        (FirebirdRouter router, address[] memory pools) = abi.decode(bridgeData, (FirebirdRouter, address[]));
```

```
    function swapExactTokensForTokens(
        IERC20TokenV06 sellToken,
        IERC20TokenV06 buyToken,
        uint amountIn,
        uint amountOutMin,
        address[] calldata path,
        address to,
        uint deadline
    ) external returns (uint[] memory amounts);
```

Then in `orders.ts` the birdge data encoder  would be `address,address[]`

Reply:
Sounds good. Handled via commit 25434c0
thx